### PR TITLE
BUG: Fix filter method so that accepts byte and unicode column names

### LINKED ISSUE #13101: BUG: .filter with unicode labels when can't encode
Edit #10506 breaks if the DataFrame contains unicode column names with non-ASCII characters.

```
import pandas as pd
df = pd.DataFrame({u'a': [1, 2, 3], u'ä': [4, 5, 6]})
df.filter(regex=u'a')
```

throws me a

```
---------------------------------------------------------------------------
UnicodeEncodeError                        Traceback (most recent call last)
<ipython-input-10-9de5a19c260e> in <module>()
----> 1 df.filter(regex=u'a')

C:\Users\...\AppData\Local\Continuum\32bit\Anaconda\envs\test\lib\site-packages\pandas\core\generic.pyc in filter(self, items, like, regex, axis)
   2013             matcher = re.compile(regex)
   2014             return self.select(lambda x: matcher.search(str(x)) is not None,
-> 2015                                axis=axis_name)
   2016         else:
   2017             raise TypeError('Must pass either `items`, `like`, or `regex`')

C:\Users\...\AppData\Local\Continuum\32bit\Anaconda\envs\test\lib\site-packages\pandas\core\generic.pyc in select(self, crit, axis)
   1545         if len(axis_values) > 0:
   1546             new_axis = axis_values[
-> 1547                 np.asarray([bool(crit(label)) for label in axis_values])]
   1548         else:
   1549             new_axis = axis_values

C:\Users\...\AppData\Local\Continuum\32bit\Anaconda\envs\test\lib\site-packages\pandas\core\generic.pyc in <lambda>(x)
   2012         elif regex:
   2013             matcher = re.compile(regex)
-> 2014             return self.select(lambda x: matcher.search(str(x)) is not None,
   2015                                axis=axis_name)
   2016         else:

UnicodeEncodeError: 'ascii' codec can't encode character u'\xe4' in position 0: ordinal not in range(128)
```

### PR DESCRIPTION
**Pull request by Licht-T**
- [x] closes #13101
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -3218,14 +3219,22 @@ def filter(self, items=None, like=None, regex=None, axis=None):
                 **{name: [r for r in items if r in labels]})
         elif like:
             def f(x):
-                if not isinstance(x, string_types):
+                if isinstance(x, binary_type):
+                    x = bytes_to_str(x)
```
Comment (jreback) [inline pandas/core/generic.py]: hmm can you codify these into a function in ``pandas/compat/__init__.py``, maybe ``to_str``, and add a doc-string so its clear what it does, put it right below the ``u/u_safe`` functions

Comment (jreback) [reply pandas/core/generic.py]: make it clear that you are turning bytes, non-strings into strings

**Review by jreback — CHANGES_REQUESTED** [review]
pls add a whatsnew for 0.21.1

File: pandas/tests/frame/test_axis_select_reindex.py [hunk 2]
```diff
@@ -884,6 +884,22 @@ def test_filter_regex_search(self):
         exp = df[[x for x in df.columns if 'BB' in x]]
         assert_frame_equal(result, exp)
 
+    @pytest.mark.parametrize('exp', ['a', u'a'])
```
Comment (jreback) [inline pandas/tests/frame/test_axis_select_reindex.py]: can you add the example from the issue, e.g. with a non-ascii character

Comment (jreback) [reply pandas/tests/frame/test_axis_select_reindex.py]: add the issue number as a comment on the tests

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by Licht-T** [comment]
Thanks @jreback, now fixed!

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @Licht-T very nice!

**Pull request merged**